pynvml compatibility

TomAugspurger opened the pull request:
Seems like some version pynvml switched from returning bytes to str (I can pin down the exact version if needed). This updates our test and type annotations to accomodate either.

jacobtomlinson reviewed (approved):
We've seen this before. There is some history and complexity around the `pynvml` and `nvidia-ml-py` dependencies here.  @rjzamora may remember better what we expect here.

But overall this PR seems to just coerce things into strings regardless of how they started, so this seems fine to me.

TomAugspurger commented:
John pointed me to https://github.com/dask/distributed/pull/8962, which did the coercion.  LMK if you and Rick have preferences on whether we should *always* try to coerce any `str | bytes` to str. Currently, I've just updated this test mock to match the new behavior (`str`) and updated the type annotations to match the pynvml version-dependent behavior.

rjzamora commented:
Thanks for taking care of this @TomAugspurger ! Given my experience with the way nvml bindings are generated in nvidia-ml-py, the safest bet is to use `ensure_unicode` (exactly like you are doing here).

jacobtomlinson commented:
Sounds good, thanks all!
